app/src/test/java/com/example/android/architecture/blueprints/todoapp/LiveDataTestUtil.kt

```diff
@@ -23,26 +22,50 @@ import java.util.concurrent.CountDownLatch
 import java.util.concurrent.TimeUnit
 import java.util.concurrent.TimeoutException
 
+/**
+ * Gets the value of a [LiveData] or waits for it to have one, with a timeout.
+ *
+ * Use this extension from host-side (JVM) tests. It's recommended to use it alongside
+ * `InstantTaskExecutorRule` or a similar mechanism to execute tasks synchronously.
+ */
 @VisibleForTesting(otherwise = VisibleForTesting.NONE)
-fun <T> LiveData<T>.awaitNextValue(
+fun <T> LiveData<T>.getOrAwaitValue(
     time: Long = 2,
-    timeUnit: TimeUnit = TimeUnit.SECONDS
+    timeUnit: TimeUnit = TimeUnit.SECONDS,
+    afterObserve: () -> Unit = {}
```

Comment:
I *think* I understand this API but not sure. Is it for testing side effects of events?

```
ld.getOrAwaitValue {
    subject.sendEventThatCausesFirstLiveDataValue()
}
```

That spelling does seem preferred to

```
subject.sendEventThatCausesFirstLiveDataValue()
ld.getOrAwaitValue()
```

Detailing the semantics in all the loopers to ensure this behavior is stable:

### InstantTaskExecutorRule:

With InstantTaskExecutor rule this will behave naturally: getOrAwaitValue will return the value created by the event.

### Robolectric
#### unPaused

This will behave identically to InstantTaskExecutorRule (I believe)

#### Paused looper
The result of `getOrAwaitValue` will be the previous value, if any. Unless the lambda progresses the shadow looper. This is "surprising", but not an atypical surprise for a paused looper.

### Android Main thread (instrumentation)
This code uses a non-volatile variable (`data`) to pass a value from the instrumentation thread to the test thread. Reading this (https://discuss.kotlinlang.org/t/non-volatile-vs-volatile-captured-refs-by-default/2122) it looks like this may not be safe in general.

Suggestion: 
 - Use a synchronization primitive around read/callback (e.g. synchronized {} )

Overall, it's important to ensure that this is returning the *last value* written to `data` and no further writes have started. The re-entry is guarded by removing the observer, but we aren't sure that the value of data read at the end is the same one written by the observer.

With that fixed:

Semantically, this will return one of 'the previous value' and 'the value produced after running this lambda' in a nondeterministic way. I'm not quite sure how to address this since it's hard to know if, or for how long, we should wait for a second value. Thoughts?

Reply:
I created so I was able to do this:

```    
        val timeTransformed = viewModel.currentTimeTransformed.getOrAwaitValue {
            // After observing, advance the clock to avoid the delay calls.
            mainCoroutineRule.advanceUntilIdle()
        }
```

Let's chat about this.

Reply:
Ah interesting, that's basically the same as the robolectric paused case.

I'm really quite stuck about how to resolve the distinction between previous and a potential next value in the real looper. I don't think it's possible with this particular API and I think it will end up nondeterministic in this situation.

Here's some radical suggestions, WDYT?

1. Remove `getOrAwaitValue()` since it may be doing too many things
2. Add 'observeForPreviousValue()` that grabs ld.value on the main thread immediately after registering the observer and uses that to ensure it never returns the `next` value. Also is there a way we could get access to the LiveData internals to keep track of the seq number (or whatever it's called)
2. Add `observeForNextValue { trigger() } that does the same trick to *discard* the previous value .
3. Add a JUnit4 Rule that provides a lifecycle, so tests can do `ld.observe(testLifecycle) { cb }`
4. Add some sort of collection result so tests can do ld<T>.captureResults(testLifecycle): Captor<T>
5. Consider providing BroadcastChannel<T> result for ld<T>.valuesAsChannel(testLifecycle

Probably more to consider, but I think that particular invocation is suggesting that this function is trying to mix captoring and execution control.